### DISCUSSION TITLE
(Blazor) WebAOT vs AOT

### DISCUSSION THREAD

**Discussion by wizardars** [dotnet/runtime#74948] [Q&A] [answered]
Why Blazor can serialize* under WebAOT w/o SourceGenerator, but AOT (console apps) can't ?

*System.Text.Json

**Comment by huoyaoyuan** [comment]
Can you confirm this is a regression with the same code, just changing `net6.0` to `net7.0` in `csproj`?

Maybe it's related to trimming and you can use the source generator approach instead.

**Reply by wizardars** [reply]
> Can you confirm this is a regression with the same code, just changing `net6.0` to `net7.0` in `csproj`?
> 
> Maybe it's related to trimming and you can use the source generator approach instead.

Yes, we have little regression .NET 7 WebAOT not working at all). https://github.com/dotnet/aspnetcore/issues/43186
But my question is not about that. I'll rewrite the title and body of the question for clarity.

**Comment by jkotas** [comment, marked as answer]
> Why Blazor can serialize* under WebAOT w/o SourceGenerator, but AOT (console apps) can't ?

In .NET 7, the console apps default to full trim mode and trim warnings by default. We believe that it is the right default configuration going forward. https://docs.microsoft.com/en-us/dotnet/core/compatibility/deployment/7.0/trim-all-assemblies has details about this breaking change.

Blazor SDK overrides the default. It uses partial trim mode and suppresses trim warnings to match how it behaved in previous versions. This configuration makes json serialization w/o source generator work in simple cases. It is a fragile configuration and it is not unusual to see it break without warning as your app gets more complex. Also, it can leave swaths of unreachable code in the application. https://github.com/dotnet/runtime/issues/74141 has an example of more complex object graph that fails to serialize in Blazor unless you are using source generators. Json source generator is the only reliable option with trimming. If you like the partial trim mode or disabled trim warnings with these caveats, you can opt into it in your .csproj file.
